### PR TITLE
dist/tools: Add script for backporting PR's

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This python script provides functionality to easily backport a merged PR to a release.

It relies on having a github API token stored locally to authenticate against github, by default it looks into ~`~/.githubtoken`~ `~/.riotgithubtoken`, but this is configurable.

The script works by fetching information from the supplied (**merged**) PR. It then looks for the latest release branch. A temporary worktree with a new branch checked out is created based on this release branch. All commit from the PR are then cherry picked into the worktree and the new branch is pushed to `origin`. It then creates a new pull request with a reference to the old PR on github and optionally puts a comment under the old PR with a reference to the new backport PR.

I haven't been able to test the last few lines of the script (creating the PR and placing the comment) as I did not want to create new backport PR's for 2018.01 :)

### Issues/PRs references

#8523

TODO:
 - [x] Make release configurable
 - [x] Copy labels to new PR
 - [x] Check if the supplied PR is already merged
 - [x] Noop mode

### CODE REVIEW

**Comment by bergzand** [comment]
Added noop mode to the todo list after some offline discussion

**Comment by kaspar030** [comment]
I tried it (and felt reckless). Some issues:
- the github token file reading needs to strip any newlines:

```
[kaspar@ng backport_pr (pr8968)]$ python3 backport_pr.py -r 2018.04 -c -d /home/kaspar/src/backport 8998
Traceback (most recent call last):
  File "backport_pr.py", line 69, in <module>
    username = g.user.get()[1]['login']
  File "/home/kaspar/.local/lib/python3.6/site-packages/agithub/base.py", line 157, in get
    return self.request('GET', url, None, headers)
  File "/home/kaspar/.local/lib/python3.6/site-packages/agithub/base.py", line 207, in request
    conn.request(method, url, requestBody.process(), headers)
  File "/usr/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1280, in _send_request
    self.putheader(hdr, value)
  File "/usr/lib/python3.6/http/client.py", line 1217, in putheader
    raise ValueError('Invalid header value %r' % (values[i],))
ValueError: Invalid header value b'Token foobar\n'
```

```./githubtoken``` contained the token + newline.


Next problem:

```
$ python3 backport_pr.py -r 2018.04 -c -d /home/kaspar/src/backport 8998
Fetching for commit: #8998: net/gcoap: increase stack size by sizeof(coap_pkt_t)
found e1f6a5afd3fd27d96a3f86921dd1d860dc7f9727 : net/gcoap: increase stack size by sizeof(coap_pkt_t)
Traceback (most recent call last):
  File "backport_pr.py", line 94, in <module>
    release_name, release_latest = get_latest_release(branches[1])
  File "backport_pr.py", line 33, in get_latest_release
    branch_name = branch['name']
TypeError: string indices must be integers
```

I stopped here for now.

This is on Arch Linux with python 3.6.

**Comment by bergzand** [comment]
Thanks for testing, will fix

**Comment by bergzand** [comment]
> Next problem:

Ah, manually selecting a branch is not really well supported yet, by default it selects the latest release by sorting them numerically :fearful:

**Comment by bergzand** [comment]
Note: The access token requires public repo access, otherwise the script fails on creating the pull request with a non-descriptive 404.
screenshot_20180423_161147

**Comment by miri64** [comment]
> Note: The access token requires public repo access, otherwise the script fails on creating the pull request with a non-descriptive 404.

Can't you catch the 404 and print a help message regarding that?

**Comment by bergzand** [comment]
> Can't you catch the 404 and print a help message regarding that?

Yes of course, I'm a bit dissapointed in that a 404 is returned and not a 403, but they probably have a good reason for that.

**Comment by miri64** [comment]
> Yes of course, I'm a bit dissapointed in that a 404 is returned and not a 403, but they probably have a good reason for that.

Most likely object obfuscation. It's in general safer (against brute force attacks) to pretend a resource does not exist than to reveal, that it does but that the requester has no permission to read it.

**Comment by cladmi** [comment]
Summary of the irl discussion on changes I would be interested in.


### Add informations on getting the token

Github->Settings->Developper settings->Personal access tokens->Generate New
Token

Only following scope is required

    repo[public repo]: Access public repositories

Name it, generate and save it to ~/.githubtoken (warning you cannot see it later anymore).

### --gitdir

 * Detect it automatically! And document/detect cases where is should be specified (as you said when using worktree)
  * -d gitdir -> say it's RIOTBASE and not the '.git' maybe name is misleading

### --comment

 * -c comment -> maybe do the comment by default and have "--no-comment" ?


### Separating steps

Instead of a "--noop", I am interested in having separate steps so I can review them:

 * create branch
     * I can verify the commit and all, maybe re-run tests also
     * I am signing these automatically created commits so would like to be able
       to review them :)
 * see the would be created pr title and message and labels
 * upload branch and create pr
 * Add a "--cleanup" option that you can run if something went wrong in the
   middle
 * And of course a lazy solution "do everything"

**Comment by cladmi** [comment]
I forgot that to make it work, I disable this line in my config

```
[push]
	default = upstream
```

It was pushing to `riot/2018-04-branch` instead of the correct branch.
I will re-test it when trying another backport.

**Comment by cladmi** [comment]
For the push default the following patch made it work:

```
git diff
diff --git a/dist/tools/backport_pr/backport_pr.py b/dist/tools/backport_pr/backport_pr.py
index 676124b..f8463bc 100644
--- a/dist/tools/backport_pr/backport_pr.py
+++ b/dist/tools/backport_pr/backport_pr.py
@@ -164,7 +164,7 @@ def main():
     # Push to github
     print("Pushing branch {} to origin".format(new_branch))
     if not args.noop:
-        repo.git.push('origin', new_branch)
+        repo.git.push('origin', '{0}:{0}'.format(new_branch))
     # Delete worktree
     print("Pruning temporary workdir at {}".format(worktree_dir))
     _delete_worktree(repo, worktree_dir)
```

**Comment by bergzand** [comment]
@cladmi Thanks for investigating!

**Comment by miri64** [comment]
Please make the script executable.

**Comment by miri64** [comment]
Some more remarks:

* some in-line doc how the keyfile is supposed to look like would be nice (maybe even generate it with help of `input()`).
* I'm pretty sure I activated the "Public Repo" access for the token, nevertheless I got the exception (maybe this was already discussed?)

**Comment by miri64** [comment]
(minor nit: I think the name `~/.githubtoken` as default token file is a little bit too generic, maybe `~/.githubbprtoken` or something so it is associated with the backport script)

**Comment by bergzand** [comment]
> (minor nit: I think the name ~/.githubtoken as default token file is a little bit too generic, maybe ~/.githubbprtoken or something so it is associated with the backport script)

If a less generic name is preferred, I'd rather name it something with riot in the name: `~/.riotgithubtoken` or something

> Please make the script executable.

Should be fixed.



> I'm pretty sure I activated the "Public Repo" access for the token, nevertheless I got the exception (maybe this was already discussed?)

Not sure why. I'd have to investigate this, the notification means that for some reason the pull request could not be created, but this could be anything.

**Comment by miri64** [comment]
> If a less generic name is preferred, I'd rather name it something with riot in the name: ~/.riotgithubtoken or something

:+1:

**Comment by cladmi** [comment]
Fixed-up the new label for "backporting-needed" https://github.com/RIOT-OS/RIOT/pull/9591#issuecomment-405883033

**Comment by cladmi** [comment]
Renamed the token file.

**Comment by miri64** [comment]
Some doc which permissions are required for the token would be nice ;-).

**Comment by miri64** [comment]
The `YYYY.MM/branch-name` naming scheme poses a problem for post-release backports, as the naming will collide with the release tag (`git` creates a file `YYYY.MM` in `.git/refs/heads/`, when the backport branch is than created it will create a directory `YYYY.MM` in `.git/refs/heads/` which name will collide with the file representing the release tag).

**Comment by jia200x** [comment]
it would be really cool to have it for the upcoming Release!

**Comment by jia200x** [comment]
> Github->Settings->Developper settings->Personal access tokens->Generate New
Token

This step can be done automatically. I usually generate Github personal tokens from python. It's even possible to tell the scope of the token there.
Does it make sense?

**Comment by miri64** [comment]
Apart from the branch naming thing: is this PR ready for review now? If yes, please squash.

**Comment by cladmi** [comment]
@miri64 updated and made it configurable. I did not tested configuring it.

**Comment by cladmi** [comment]
> Why from the environment and not the command-line arguments oO. That really adds some inconsistencies.

I guess I was not really thinking about the configuration for this one. I focused more on the `change name`. I will update it.

**Comment by aabadie** [comment]
Is this still WIP ? It doesn't seem so. It would be great to have this merged before next feature freeze.

**Comment by cladmi** [comment]
I can squash if you want.

**Comment by aabadie** [comment]
ping @cladmi

**Comment by cladmi** [comment]
In progress.

**Comment by cladmi** [comment]
I did two push force. The first one being rebase and then the squash.
I used the beginning of the README as commit message.

I inlined the `usage` update as can be seen on the last force-pushed comparison.

**Comment by cladmi** [comment]
I contributed a lot to this script so do not feel in a position where I should ACK my own code.

**Comment by miri64** [comment]
I used the script quite recently to create #10757, so I can confirm it works.

**Comment by miri64** [comment]
Gave the code-style badge as well, since it is checked by `flake8` ;-)

**Comment by miri64** [comment]
I'll look into the code now, but I think we can go ahead with feature freeze. The script is not required to be in the release branch to create a backport ;-).

File: dist/tools/backport_pr/backport_pr.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+                             "push and create the PR")
+    parser.add_argument("-r", "--release-branch", type=str,
+                        help="Base the backport on this branch, "
+                             "default is the latest")
+    parser.add_argument("--backport-branch-fmt", type=str,
+                        default=BACKPORT_BRANCH,
+                        help="Backport branch format. "
+                             "Fields '{release}' and '{origbranch} will be "
+                             "replaced by the release name and remote branch "
+                             "name.")
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: default doc missing

File: dist/tools/backport_pr/backport_pr.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    parser.add_argument('-d', '--gitdir', type=str, default=os.getcwd(),
+                        help="Base git repo to work from")
+    parser.add_argument("PR", type=int, help="Pull request number to backport")
+    args = parser.parse_args()
+
+    gittoken = args.keyfile.read().strip()
+    g = GitHub(token=gittoken)
+    # TODO: exception handling
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: I guess this qualifies the script still to be WIP? ;-)

File: dist/tools/backport_pr/backport_pr.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    status, user = g.user.get()
+    if status != 200:
+        print("Could not retrieve user: {}".format(user['message']))
+        exit(1)
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: This

```py
status, obj = endpoint.get()
if status != 200:
   print("error")
   exit(val)
```

pattern shows up quite often. Any chance to put this into its own function?

File: dist/tools/backport_pr/backport_pr.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    status, commits = g.repos[ORG][REPO].pulls[args.PR].commits.get()
+    if status != 200:
+        print("No commits found for #{}: {}".format(args.PR,
+                                                    commits['message']))
+        sys.exit(3)
+    for commit in commits:
+        print("found {} : {}".format(commit['sha'],
+                                     commit['commit']['message']))
+
+    # Find latest release branch
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: General rule of thumb for me: If you can summarize a large chunk of code with a comment above it: Put it in its own function ;-).

File: dist/tools/backport_pr/backport_pr.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                          branches['message']))
+            sys.exit(4)
+        release_shortname, release_fullname = _get_latest_release(branches)
+        if not release_fullname:
+            print("No release branch found, exiting")
+            sys.exit(5)
+    print("Backport based on branch {}".format(release_fullname))
+
+    repo = git.Repo(args.gitdir)
+    # Fetch current upstream
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: Dito

File: dist/tools/backport_pr/backport_pr.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    upstream_remote = _get_upstream(repo)
+    if not upstream_remote:
+        print("No upstream remote found, can't fetch")
+        exit(6)
+    print("Fetching {} remote".format(upstream_remote))
+
+    upstream_remote.fetch()
+    # Build topic branch in temp dir
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: Dito

File: dist/tools/backport_pr/backport_pr.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    if not args.noop:
+        repo.git.push('origin', '{0}:{0}'.format(new_branch))
+    # Delete worktree
+    print("Pruning temporary workdir at {}".format(worktree_dir))
+    _delete_worktree(repo, worktree_dir)
+
+    labels = _get_labels(pulldata)
+    merger = pulldata['merged_by']['login']
+    if not args.noop:
+        # Open new PR on github
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: And so on...

**Comment by aabadie** [comment]
ok let's wait, it's not mandatory to have it in the release.

File: dist/tools/backport_pr/backport_pr.py (hunk 1 above)
Comment (cladmi) [reply dist/tools/backport_pr/backport_pr.py]: I could do it with 
``` diff
-    parser = argparse.ArgumentParser()
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
```
But then other default do not look that great :/

**Comment by cladmi** [comment]
If we start looking into the script quality, I would also point the `pylint` output:

```
pylint ./dist/tools/backport_pr/backport_pr.py
************* Module backport_pr
dist/tools/backport_pr/backport_pr.py:111:6: W0511: TODO: exception handling (fixme)
dist/tools/backport_pr/backport_pr.py:187:0: C0330: Wrong hanging indentation (remove 4 spaces).
                'title': "{} [backport {}]".format(pulldata['title'],
            |   ^ (bad-continuation)
dist/tools/backport_pr/backport_pr.py:189:0: C0330: Wrong hanging indentation (remove 4 spaces).
                'head': '{}:{}'.format(username, new_branch),
            |   ^ (bad-continuation)
dist/tools/backport_pr/backport_pr.py:190:0: C0330: Wrong hanging indentation (remove 4 spaces).
                'base': release_fullname,
            |   ^ (bad-continuation)
dist/tools/backport_pr/backport_pr.py:191:0: C0330: Wrong hanging indentation (remove 4 spaces).
                'body': "# Backport of #{}\n\n{}".format(args.PR,
            |   ^ (bad-continuation)
dist/tools/backport_pr/backport_pr.py:193:0: C0330: Wrong hanging indentation (remove 4 spaces).
                'maintainer_can_modify': True
            |   ^ (bad-continuation)
dist/tools/backport_pr/backport_pr.py:1:0: C0111: Missing module docstring (missing-docstring)
dist/tools/backport_pr/backport_pr.py:12:0: E0401: Unable to import 'agithub.GitHub' (import-error)
dist/tools/backport_pr/backport_pr.py:34:0: C0103: Argument name "pr" doesn't conform to snake_case naming style (invalid-name)
dist/tools/backport_pr/backport_pr.py:45:11: C1801: Do not use `len(SEQUENCE)` to determine if a sequence is empty (len-as-condition)
dist/tools/backport_pr/backport_pr.py:47:11: C1801: Do not use `len(SEQUENCE)` to determine if a sequence is empty (len-as-condition)
dist/tools/backport_pr/backport_pr.py:70:0: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
dist/tools/backport_pr/backport_pr.py:82:0: C0111: Missing function docstring (missing-docstring)
dist/tools/backport_pr/backport_pr.py:82:0: R0914: Too many local variables (28/15) (too-many-locals)
dist/tools/backport_pr/backport_pr.py:110:4: C0103: Variable name "g" doesn't conform to snake_case naming style (invalid-name)
dist/tools/backport_pr/backport_pr.py:186:8: C0103: Variable name "pr" doesn't conform to snake_case naming style (invalid-name)
dist/tools/backport_pr/backport_pr.py:82:0: R0912: Too many branches (16/12) (too-many-branches)
dist/tools/backport_pr/backport_pr.py:82:0: R0915: Too many statements (80/50) (too-many-statements)
dist/tools/backport_pr/backport_pr.py:14:0: C0411: standard import "import argparse" should be placed before "from agithub.GitHub import GitHub" (wrong-import-order)
dist/tools/backport_pr/backport_pr.py:15:0: C0411: standard import "import os" should be placed before "from agithub.GitHub import GitHub" (wrong-import-order)
dist/tools/backport_pr/backport_pr.py:16:0: C0411: standard import "import os.path" should be placed before "from agithub.GitHub import GitHub" (wrong-import-order)
dist/tools/backport_pr/backport_pr.py:17:0: C0411: standard import "import shutil" should be placed before "from agithub.GitHub import GitHub" (wrong-import-order)
dist/tools/backport_pr/backport_pr.py:18:0: C0411: standard import "import sys" should be placed before "from agithub.GitHub import GitHub" (wrong-import-order)

------------------------------------------------------------------
Your code has been rated at 8.00/10 (previous run: 8.00/10, +0.00)

```

And enable the `--max-complexity` to `flake8`. I have seen other people use a value of `10` as default. 
```
flake8 --max-complexity=8 ./dist/tools/backport_pr/backport_pr.py
./dist/tools/backport_pr/backport_pr.py:82:1: C901 'main' is too complex (16)
```

So we could still refine the script I guess.

File: dist/tools/backport_pr/backport_pr.py (hunk 1 above)
Comment (miri64) [reply dist/tools/backport_pr/backport_pr.py]: How about doing it just like the other defaults for now. Other changes can always come later ;-).

**Comment by miri64** [comment]
Why rebase at all if there is no conflict ;-).

**Comment by cladmi** [comment]
I wanted to be based on the branch where `compile_and_test_for_board` was included to take the `tox.ini` file and have it as reference.

File: dist/tools/backport_pr/backport_pr.py (hunk 3 above)
Comment (cladmi) [reply dist/tools/backport_pr/backport_pr.py]: That should just raise an exception.

**Comment by miri64** [comment]
> I wanted to be based on the branch where `compile_and_test_for_board` was included to take the `tox.ini` file and have it as reference.

[Tox](https://tox.chat/)?

**Comment by cladmi** [comment]
https://tox.readthedocs.io

File: dist/tools/backport_pr/backport_pr.py (hunk 3 above)
Comment (miri64) [reply dist/tools/backport_pr/backport_pr.py]: The function should raise an exception? Yes. Otherwise, I'm not sure what you mean.

Comment (cladmi) [reply dist/tools/backport_pr/backport_pr.py]: Yes that is what I meant indeed.

**Comment by cladmi** [comment]
Before somebody gets crazy when refactoring into functions the `Github` object behaves quick strangely and should alway be used with the full `g.repos[ORG][REPO].pulls[args.PR]` syntax. If you re-use a `g.repos[ORG][REPO]` object, things gets complicated as doing `pulls[]` does modify the original object. (oh don't ask why, please don't ask why)

**Comment by bergzand** [comment]
> Before somebody gets crazy when refactoring into functions the `Github` object behaves quick strangely and should alway be used with the full `g.repos[ORG][REPO].pulls[args.PR]` syntax. If you re-use a `g.repos[ORG][REPO]` object, things gets complicated as doing `pulls[]` does modify the original object. (oh don't ask why, please don't ask why)

I've also noticed this once. The whole `Github` object (or rather the agithub module) is a weird automatic API object that *generates* API requests based on how you access the object.

**Comment by cladmi** [comment]
> I've also noticed this once. The whole Github object (or rather the agithub module) is a weird automatic API object that generates API requests based on how you access the object.

Why would not a getter be allowed to modify an object? I think it is a really nice design pattern ><

File: dist/tools/backport_pr/backport_pr.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        exit(1)
+    username = user['login']
+    status, pulldata = g.repos[ORG][REPO].pulls[args.PR].get()
+    if status != 200:
+        print("Commit #{} not found: {}".format(args.PR, pulldata['message']))
+        sys.exit(2)
+    if not pulldata['merged']:
+        print("Original PR not yet merged")
+        exit(0)
+    print("Fetching for commit: #{}: {}".format(args.PR, pulldata['title']))
```
Comment (miri64) [inline dist/tools/backport_pr/backport_pr.py]: One thing I noticed here: if I have a multi-line commit message it shows the whole message, not just the summary line.

**Comment by miri64** [comment]
Ping for squashing btw.

**Comment by cladmi** [comment]
I squashed and rebased. I in-lined a fix for `Wrong hanging indentation` warnings.

The `tox` script is still showing errors but gives things that still need to be improved.

**Comment by miri64** [comment]
Ok, let's fix those in follow-ups.

**Comment by bergzand** [comment]
@cladmi Thank you for managing this PR!

**Comment by cladmi** [comment]
@bergzand you wrote all the logic, I only handled the fiddling which was easier :)

**Comment by cladmi** [comment]
So thanks to you !

**Pull request merged**